File: Python/excavatorModel.py
import casadi as csd
from enum import Enum
import excavatorConstants as C

def forwardKinematics(q):
    alpha = q[0]
    beta = q[1]
    gamma = q[2]

    x = C.lenBA*csd.cos(alpha) + C.lenAL*csd.cos(alpha + beta) + C.lenLM*csd.cos(alpha + beta + gamma)
    y = C.lenBA*csd.sin(alpha) + C.lenAL*csd.sin(alpha + beta) + C.lenLM*csd.sin(alpha + beta + gamma)
    theta = alpha + beta + gamma
    return csd.vertcat(x, y, theta)

def inverseKinematics(pose):
    xTip = pose[0]
    yTip = pose[1]
    thetaTip = pose[2]

    xJointBucket = xTip - C.lenLM*csd.cos(thetaTip)
    yJointBucket = yTip - C.lenLM*csd.sin(thetaTip)

    cosBeta = (xJointBucket**2 + yJointBucket**2 - C.lenBA**2 - C.lenAL**2)/(2*C.lenBA*C.lenAL)
    sinBeta = -csd.sqrt(1 - cosBeta**2)

    sinAlpha = ((C.lenBA + C.lenAL*cosBeta)*yJointBucket - C.lenAL*sinBeta*xJointBucket)/(xJointBucket**2 + yJointBucket**2)
    cosAlpha = ((C.lenBA + C.lenAL*cosBeta)*xJointBucket + C.lenAL*sinBeta*yJointBucket)/(xJointBucket**2 + yJointBucket**2)

    alpha = csd.atan2(sinAlpha, cosAlpha)
    beta = csd.atan2(sinBeta, cosBeta)
    gamma = thetaTip - alpha - beta

    return csd.vertcat(alpha, beta, gamma)

def inverseDynamics(q, qDot, qDDot, F):
    alpha = q[0]
    beta = q[1]
    gamma = q[2]
    alphaDot = qDot[0]
    betaDot = qDot[1]
    gammaDot = qDot[2]

    jacBoom = csd.vertcat(csd.horzcat(-C.lenBCoMBoom*csd.sin(alpha + C.angABCoMBoom), 0, 0),
                          csd.horzcat(C.lenBCoMBoom*csd.cos(alpha + C.angABCoMBoom), 0, 0),
                          csd.horzcat(1, 0, 0))
    jacArm = csd.vertcat(csd.horzcat(-C.lenBA*csd.sin(alpha) - C.lenACoMArm*csd.sin(alpha + beta + C.angLACoMArm),
                                     -C.lenACoMArm*csd.sin(alpha + beta + C.angLACoMArm),
                                     0),
                         csd.horzcat(C.lenBA*csd.cos(alpha) + C.lenACoMArm*csd.cos(alpha + beta + C.angLACoMArm),
                                     C.lenACoMArm*csd.cos(alpha + beta + C.angLACoMArm),
                                     0),
                         csd.horzcat(1, 1, 0))
    jacBucket = csd.vertcat(csd.horzcat(-C.lenBA*csd.sin(alpha) - C.lenAL*csd.sin(alpha + beta) - C.lenLCoMBucket*csd.sin(alpha + beta + gamma + C.angMLCoMBucket),
                                        -C.lenAL*csd.sin(alpha + beta) - C.lenLCoMBucket*csd.sin(alpha + beta + gamma + C.angMLCoMBucket),
                                        -C.lenLCoMBucket*csd.sin(alpha + beta + gamma + C.angMLCoMBucket)),
                            csd.horzcat(C.lenBA*csd.cos(alpha) + C.lenAL*csd.cos(alpha + beta) + C.lenLCoMBucket*csd.cos(alpha + beta + gamma + C.angMLCoMBucket),
                                        C.lenAL*csd.cos(alpha + beta) + C.lenLCoMBucket*csd.cos(alpha + beta + gamma + C.angMLCoMBucket),
                                        C.lenLCoMBucket*csd.cos(alpha + beta + gamma + C.angMLCoMBucket)),
                            csd.horzcat(1, 1, 1))
    jacBoomDot = csd.vertcat(csd.horzcat(-C.lenBCoMBoom*csd.cos(alpha + C.angABCoMBoom)*alphaDot, 0, 0),
                             csd.horzcat(-C.lenBCoMBoom*csd.sin(alpha + C.angABCoMBoom)*alphaDot, 0, 0),
                             csd.horzcat(0, 0, 0))
    jacArmDot = csd.vertcat(csd.horzcat(-C.lenBA*csd.cos(alpha)*alphaDot - C.lenACoMArm*csd.cos(alpha + beta + C.angLACoMArm)*(alphaDot + betaDot),
                                        -C.lenACoMArm*csd.cos(alpha + beta + C.angLACoMArm)*(alphaDot + betaDot),
                                        0),
                            csd.horzcat(-C.lenBA*csd.sin(alpha)*alphaDot - C.lenACoMArm*csd.sin(alpha + beta + C.angLACoMArm)*(alphaDot + betaDot),
                                        -C.lenACoMArm*csd.sin(alpha + beta + C.angLACoMArm)*(alphaDot + betaDot),
                                        0),
                            csd.horzcat(0, 0, 0))
    jacBucketDot = csd.vertcat(csd.horzcat(-C.lenBA*csd.cos(alpha)*alphaDot - C.lenAL*csd.cos(alpha + beta)*(alphaDot + betaDot) - C.lenLCoMBucket*csd.cos(alpha + beta + gamma + C.angMLCoMBucket)*(alphaDot + betaDot + gammaDot),
                                           -C.lenAL*csd.cos(alpha + beta)*(alphaDot + betaDot) - C.lenLCoMBucket*csd.cos(alpha + beta + gamma + C.angMLCoMBucket)*(alphaDot + betaDot + gammaDot),
                                           -C.lenLCoMBucket*csd.cos(alpha + beta + gamma + C.angMLCoMBucket)*(alphaDot + betaDot + gammaDot)),
                               csd.horzcat(-C.lenBA*csd.sin(alpha)*alphaDot - C.lenAL*csd.sin(alpha + beta)*(alphaDot + betaDot) - C.lenLCoMBucket*csd.sin(alpha + beta + gamma + C.angMLCoMBucket)*(alphaDot + betaDot + gammaDot),
                                           -C.lenAL*csd.sin(alpha + beta)*(alphaDot + betaDot) - C.lenLCoMBucket*csd.sin(alpha + beta + gamma + C.angMLCoMBucket)*(alphaDot + betaDot + gammaDot),
                                           -C.lenLCoMBucket*csd.sin(alpha + beta + gamma + C.angMLCoMBucket)*(alphaDot + betaDot + gammaDot)),
                               csd.horzcat(0, 0, 0))

    M = (csd.transpose(jacBoom[0:2, :])@C.massBoom@jacBoom[0:2, :] + csd.transpose(jacBoom[2, :])@C.moiBoom@jacBoom[2, :] +
         csd.transpose(jacArm[0:2, :])@C.massArm@jacArm[0:2, :] + csd.transpose(jacArm[2, :])@C.moiArm@jacArm[2, :] +
         csd.transpose(jacBucket[0:2, :])@C.massBucket@jacBucket[0:2, :] + csd.transpose(jacBucket[2, :])@C.moiBucket@jacBucket[2, :])
    b = (csd.transpose(jacBoom[0:2, :])@C.massBoom@jacBoomDot[0:2, :]@qDot + csd.transpose(jacBoom[2, :])@(C.moiBoom@jacBoomDot[2, :]@qDot) +
         csd.transpose(jacArm[0:2, :])@C.massArm@jacArmDot[0:2, :]@qDot + csd.transpose(jacArm[2, :])@(C.moiArm@jacArmDot[2, :]@qDot) +
         csd.transpose(jacBucket[0:2, :])@C.massBucket@jacBucketDot[0:2, :]@qDot + csd.transpose(jacBucket[2, :])@(C.moiBucket@jacBucketDot[2, :]@qDot))
    g = (-csd.transpose(jacBoom[0:2, :])@C.massBoom@C.g -
         csd.transpose(jacArm[0:2, :])@C.massArm@C.g -
         csd.transpose(jacBucket[0:2, :])@C.massBucket@C.g)
    extF = csd.transpose(jacBucket[0:2, :])@F

    return M@qDDot + b + g - extF

def actuatorLen(q):
    alpha = q[0]
    beta = q[1]
    gamma = q[2]

    # R = C.lenBC
    # theta = csd.atan2(C.iBC[1], C.iBC[0])
    # lenBoom = csd.sqrt(-2*R*C.lenBD*csd.cos(alpha - theta + C.angABD) + C.lenBD**2 + C.lenBC**2)
    lenBoom = 0.0086*alpha**4 - 0.0459*alpha**3 - 0.0104*alpha**2 + 0.2956*alpha + 1.042

    # R = C.lenAE
    # theta = csd.atan2(C.bAE[0], C.bAE[1])
    # lenArm = csd.sqrt(-2*R*C.lenAF*csd.sin(beta + theta + C.angFAL) + C.lenAF**2 + C.lenAE**2)
    lenArm = 0.0078*beta**4 + 0.0917*beta**3 + 0.2910*beta**2 + 0.0646*beta + 1.0149

    lenBucket = 0.0048*gamma**4 + 0.0288*gamma**3 + 0.0225*gamma**2 - 0.1695*gamma + 0.9434

    return csd.vertcat(lenBoom, lenArm, lenBucket)

def actuatorVelFactor(q):
    alpha = q[0]
    beta = q[1]
    gamma = q[2]

    velFactorBoom = 0.0344*alpha**3 - 0.1377*alpha**2 - 0.0208*alpha + 0.2956
    velFactorArm = 0.0312*beta**3 + 0.2751*beta**2 + 0.582*beta + 0.0646
    velFactorBucket = 0.0192*gamma**3 + 0.0864*gamma**2 + 0.045*gamma - 0.1695

    return csd.vertcat(velFactorBoom, velFactorArm, velFactorBucket)

def actuatorVel(q, qDot):
    alpha = q[0]
    beta = q[1]
    gamma = q[2]
    alphaDot = qDot[0]
    betaDot = qDot[1]
    gammaDot = qDot[2]

    # len = actuatorLen(q)
    # lenBoom = len[0]
    # lenArm = len[1]
    # lenBucket = len[2]

    # angBDC = csd.acos((lenBoom**2 + C.lenBD**2 - C.lenBC**2)/(2*lenBoom*C.lenBD))
    # theta = csd.pi/2 - angBDC
    # lenBoomDot = alphaDot*C.lenBD*csd.cos(theta)
    lenBoomDot = alphaDot*(0.0344*alpha**3 - 0.1377*alpha**2 - 0.0208*alpha + 0.2956)

    # angAFE = csd.acos((lenArm**2 + C.lenAF**2 - C.lenAE**2)/(2*lenArm*C.lenAF))
    # theta = angAFE - csd.pi/2
    # lenArmDot = -betaDot*C.lenAF*csd.cos(theta)
    lenArmDot = betaDot*(0.0312*beta**3 + 0.2751*beta**2 + 0.582*beta + 0.0646)

    lenBucketDot = gammaDot*(0.0192*gamma**3 + 0.0864*gamma**2 + 0.045*gamma - 0.1695)

    return csd.vertcat(lenBoomDot, lenArmDot, lenBucketDot)

def motorVel(q, qDot):
    lenDot = actuatorVel(q, qDot)
    lenBoomDot = lenDot[0]
    lenArmDot = lenDot[1]
    lenBucketDot = lenDot[2]

    angVelMotorBoom = 2444.16*lenBoomDot
    angVelMotorArm = 2444.16*lenArmDot
    angVelMotorBucket = 2444.16*lenBucketDot

    return csd.vertcat(angVelMotorBoom, angVelMotorArm, angVelMotorBucket)

def motorTorque(q, qDot, qDDot, F):
    T = inverseDynamics(q, qDot, qDDot, F)
    TBoom = T[0]
    TArm = T[1]
    TBucket = T[2]

    r = actuatorVelFactor(q)
    rBoom = r[0]
    rArm = r[1]
    rBucket = r[2]

    TMotorBoom = TBoom/(2444.16*rBoom*0.64)
    TMotorArm = TArm/(2444.16*rArm*0.64)
    TMotorBucket = TBucket/(2444.16*rBucket*0.64)

    return csd.vertcat(TMotorBoom, TMotorArm, TMotorBucket)

class DutyCycle(Enum):
    S1 = 0
    S2_60 = 1
    S2_30 = 2
    PEAK = 3

def motorTorqueLimit(motorVel, dutyCycle):
    match dutyCycle:
        case DutyCycle.S1:
            return -1.4073e-7*motorVel**3 + 1.7961e-5*motorVel**2 - 0.0147*motorVel + 19.9091
        case DutyCycle.S2_60:
            return -1.3568e-10*motorVel**4 -1.4682e-7*motorVel**3 + 5.5744e-5*motorVel**2 - 0.0159*motorVel + 25.0769
        case DutyCycle.S2_30:
            return -2.0433e-7*motorVel**3 + 5.1865e-5*motorVel**2 - 0.0105*motorVel + 30.9119
        case DutyCycle.PEAK:
            return 4.0269e-9*motorVel**4 - 3.7090e-6*motorVel**3 + 8.513e-4*motorVel**2 - 0.05787*motorVel + 60.2614
        case _:
            return -1.4073e-7*motorVel**3 + 1.7961e-5*motorVel**2 - 0.0147*motorVel + 19.9091

File: MATLAB/MATLAB/excavatorModel.m
classdef excavatorModel < articulatedArmModel
    properties
        c; %excavatorConstants
        lenBoom;
        lenArm;
        lenBucket;
    end

    methods
        % Constructor
        function obj = excavatorModel(excavatorConstants, alpha, beta, gamma)
            obj = obj@articulatedArmModel(excavatorConstants.lenBA, ...
                excavatorConstants.lenAL, ...
                excavatorConstants.lenLM, ...
                alpha, beta, gamma);
            obj.c = excavatorConstants;
        end
        
        %{
        Set the boom, arm and bucket actuator lengths and update the
        corresponding angles
        Input: Boom, arm and bucket actuator lengths
        Output: Boom, arm and bucket angles
        %}
        function [alpha, beta, gamma] = setLengths(obj, lenBoom, lenArm, lenBucket)
            obj.lenBoom = lenBoom;
            obj.lenArm = lenArm;
            obj.lenBucket = lenBucket;
            
            % Boom (0.905 - 1.305 m)
            R = obj.c.lenBC; % R formula (cos version)
            theta = atan2(obj.c.iBC(2), obj.c.iBC(1));
            obj.ang1 = acos((-lenBoom^2 + obj.c.lenBD^2 + obj.c.lenBC^2) ...
                /(2*R*obj.c.lenBD)) + theta - obj.c.angABD;
            alpha = obj.ang1;

            % Arm (1.046 - 1.556 m)
            R = obj.c.lenAE; % R formula (sin version)
            theta = atan2(obj.c.bAE(1), obj.c.bAE(2));
            obj.ang2 = asin((-lenArm^2 + obj.c.lenAF^2 + obj.c.lenAE^2) ...
                /(2*R*obj.c.lenAF)) - theta - obj.c.angFAL;
            beta = obj.ang2;

            % Bucket (0.84 - 1.26 m)
            R = obj.c.lenJG; % R formula (sin version)
            theta = atan2(obj.c.aJG(1), obj.c.aJG(2));
            angLJH = asin((-lenBucket^2 + obj.c.lenHJ^2 + obj.c.lenJG^2) ...
                /(2*R*obj.c.lenHJ)) - theta;

            R = sqrt((obj.c.lenJL - obj.c.lenHJ*cos(angLJH))^2 + (obj.c.lenHJ*sin(angLJH))^2); % R formula (cos version)
            theta = atan2(obj.c.lenHJ*sin(angLJH), obj.c.lenJL - obj.c.lenHJ*cos(angLJH));
            obj.ang3 = acos((obj.c.lenHK^2 - obj.c.lenJL^2 - obj.c.lenLK^2 - obj.c.lenHJ^2 + ...
                2*obj.c.lenJL*obj.c.lenHJ*cos(angLJH))/(2*R*obj.c.lenLK)) - ...
                theta - obj.c.angKLM;
            gamma = obj.ang3;
        end
        
        %{
        Set the boom, arm and bucket angles and update the
        corresponding actuator lengths
        Input: Boom, arm and bucket angles
        Output: Boom, arm and bucket actuator lengths
        %}        
        function [lenBoom, lenArm, lenBucket] = setAngles(obj, alpha, beta, gamma)
            obj.ang1 = alpha;
            obj.ang2 = beta;
            obj.ang3 = gamma;

            % Boom
            R = obj.c.lenBC; % R formula (cos version)
            theta = atan2(obj.c.iBC(2), obj.c.iBC(1));
            obj.lenBoom = sqrt(-2*R*obj.c.lenBD*cos(alpha - theta + obj.c.angABD) + obj.c.lenBD^2 + obj.c.lenBC^2);
            lenBoom = obj.lenBoom;

            % Arm
            R = obj.c.lenAE; % R formula (sin version)
            theta = atan2(obj.c.bAE(1), obj.c.bAE(2)); 
            obj.lenArm = sqrt(-2*R*obj.c.lenAF*sin(beta + theta + obj.c.angFAL) + obj.c.lenAF^2 + obj.c.lenAE^2);
            lenArm = obj.lenArm;

            %Bucket
            aAK = obj.c.aAL + obj.c.lenLK*[cos(gamma + obj.c.angKLM); sin(gamma + obj.c.angKLM)];
            aJK = aAK - obj.c.aAJ;
            lenJK = norm(aJK);
            cosAngHJK = (lenJK^2 + obj.c.lenHJ^2 - obj.c.lenHK^2)/(2*lenJK*obj.c.lenHJ);
            sinAngHJK = sqrt(1 - cosAngHJK^2);
            aAH = obj.c.aAJ + (obj.c.lenHJ/lenJK)*[cosAngHJK -sinAngHJK; sinAngHJK cosAngHJK]*aJK;
            obj.lenBucket = sqrt((aAH(1) - obj.c.aAG(1))^2 + (aAH(2) - obj.c.aAG(2))^2);
            lenBucket = obj.lenBucket;
        end

        %{
        Calculate boom, arm and bucket virtual joint torques
        Input: Boom, arm and bucket angular accelerations and velocities
        Output: Boom, arm and bucket virtual joint torques
        %}
        function [TBoom, TArm, TBucket] = inverseDynamics(obj, ang1Dot, ang2Dot, ang3Dot, ang1DDot, ang2DDot, ang3DDot, F)
            qDot = [ang1Dot; ang2Dot; ang3Dot];
            qDDot = [ang1DDot; ang2DDot; ang3DDot];
            
            % Jacobians of the centre of masses
            iJacBoom = [-obj.c.lenBCoMBoom*sin(obj.ang1 + obj.c.angABCoMBoom), 0, 0; ...
                obj.c.lenBCoMBoom*cos(obj.ang1 + obj.c.angABCoMBoom), 0, 0; ...
                1, 0, 0];
            iJacArm = [-obj.c.lenBA*sin(obj.ang1) - obj.c.lenACoMArm*sin(obj.ang1 + obj.ang2 + obj.c.angLACoMArm), ...
                -obj.c.lenACoMArm*sin(obj.ang1 + obj.ang2 + obj.c.angLACoMArm), ...
                0; ...
                obj.c.lenBA*cos(obj.ang1) + obj.c.lenACoMArm*cos(obj.ang1 + obj.ang2 + obj.c.angLACoMArm), ...
                obj.c.lenACoMArm*cos(obj.ang1 + obj.ang2 + obj.c.angLACoMArm), ...
                0; ...
                1, 1, 0];
            iJacBucket = [-obj.c.lenBA*sin(obj.ang1) - obj.c.lenAL*sin(obj.ang1 + obj.ang2) - obj.c.lenLCoMBucket*sin(obj.ang1 + obj.ang2 + obj.ang3 + obj.c.angMLCoMBucket), ...
                -obj.c.lenAL*sin(obj.ang1 + obj.ang2) - obj.c.lenLCoMBucket*sin(obj.ang1 + obj.ang2 + obj.ang3 + obj.c.angMLCoMBucket), ...
                -obj.c.lenLCoMBucket*sin(obj.ang1 + obj.ang2 + obj.ang3 + obj.c.angMLCoMBucket); ...
                obj.c.lenBA*cos(obj.ang1) + obj.c.lenAL*cos(obj.ang1 + obj.ang2) + obj.c.lenLCoMBucket*cos(obj.ang1 + obj.ang2 + obj.ang3 + obj.c.angMLCoMBucket), ...
                obj.c.lenAL*cos(obj.ang1 + obj.ang2) + obj.c.lenLCoMBucket*cos(obj.ang1 + obj.ang2 + obj.ang3 + obj.c.angMLCoMBucket), ...
                obj.c.lenLCoMBucket*cos(obj.ang1 + obj.ang2 + obj.ang3 + obj.c.angMLCoMBucket); ...
                1, 1, 1];
            iJacBoomDot = [-obj.c.lenBCoMBoom*cos(obj.ang1 + obj.c.angABCoMBoom)*ang1Dot, 0, 0; ...
                -obj.c.lenBCoMBoom*sin(obj.ang1 + obj.c.angABCoMBoom)*ang1Dot, 0, 0; ...
                0, 0, 0];
            iJacArmDot = [-obj.c.lenBA*cos(obj.ang1)*ang1Dot - obj.c.lenACoMArm*cos(obj.ang1 + obj.ang2 + obj.c.angLACoMArm)*(ang1Dot + ang2Dot), ...
                -obj.c.lenACoMArm*cos(obj.ang1 + obj.ang2 + obj.c.angLACoMArm)*(ang1Dot + ang2Dot), ...
                0; ...
                -obj.c.lenBA*sin(obj.ang1)*ang1Dot - obj.c.lenACoMArm*sin(obj.ang1 + obj.ang2 + obj.c.angLACoMArm)*(ang1Dot + ang2Dot), ...
                -obj.c.lenACoMArm*sin(obj.ang1 + obj.ang2 + obj.c.angLACoMArm)*(ang1Dot + ang2Dot), ...
                0; ...
                0, 0, 0];
            iJacBucketDot = [-obj.c.lenBA*cos(obj.ang1)*ang1Dot - obj.c.lenAL*cos(obj.ang1 + obj.ang2)*(ang1Dot + ang2Dot) - obj.c.lenLCoMBucket*cos(obj.ang1 + obj.ang2 + obj.ang3 + obj.c.angMLCoMBucket)*(ang1Dot + ang2Dot + ang3Dot), ...
                -obj.c.lenAL*cos(obj.ang1 + obj.ang2)*(ang1Dot + ang2Dot) - obj.c.lenLCoMBucket*cos(obj.ang1 + obj.ang2 + obj.ang3 + obj.c.angMLCoMBucket)*(ang1Dot + ang2Dot + ang3Dot), ...
                -obj.c.lenLCoMBucket*cos(obj.ang1 + obj.ang2 + obj.ang3 + obj.c.angMLCoMBucket)*(ang1Dot + ang2Dot + ang3Dot); ...
                -obj.c.lenBA*sin(obj.ang1)*ang1Dot - obj.c.lenAL*sin(obj.ang1 + obj.ang2)*(ang1Dot + ang2Dot) - obj.c.lenLCoMBucket*sin(obj.ang1 + obj.ang2 + obj.ang3 + obj.c.angMLCoMBucket)*(ang1Dot + ang2Dot + ang3Dot), ...
                -obj.c.lenAL*sin(obj.ang1 + obj.ang2)*(ang1Dot + ang2Dot) - obj.c.lenLCoMBucket*sin(obj.ang1 + obj.ang2 + obj.ang3 + obj.c.angMLCoMBucket)*(ang1Dot + ang2Dot + ang3Dot), ...
                -obj.c.lenLCoMBucket*sin(obj.ang1 + obj.ang2 + obj.ang3 + obj.c.angMLCoMBucket)*(ang1Dot + ang2Dot + ang3Dot); ...
                0, 0, 0];
            iJacM = [-obj.c.lenBA*sin(obj.ang1) - obj.c.lenAL*sin(obj.ang1 + obj.ang2) - obj.c.lenLM*sin(obj.ang1 + obj.ang2 + obj.ang3), ...
                -obj.c.lenAL*sin(obj.ang1 + obj.ang2) - obj.c.lenLM*sin(obj.ang1 + obj.ang2 + obj.ang3), ...
                -obj.c.lenLM*sin(obj.ang1 + obj.ang2 + obj.ang3); ...
                obj.c.lenBA*cos(obj.ang1) + obj.c.lenAL*cos(obj.ang1 + obj.ang2) + obj.c.lenLM*cos(obj.ang1 + obj.ang2 + obj.ang3), ...
                obj.c.lenAL*cos(obj.ang1 + obj.ang2) + obj.c.lenLM*cos(obj.ang1 + obj.ang2 + obj.ang3), ...
                obj.c.lenLM*cos(obj.ang1 + obj.ang2 + obj.ang3); ...
                1, 1, 1];
            
            % Mass matrix
            M = iJacBoom(1:2,:)'*obj.c.massBoom*iJacBoom(1:2,:) + iJacBoom(3,:)'*obj.c.moiBoom*iJacBoom(3,:) + ...
                iJacArm(1:2,:)'*obj.c.massArm*iJacArm(1:2,:) + iJacArm(3,:)'*obj.c.moiArm*iJacArm(3,:) + ...
                iJacBucket(1:2,:)'*obj.c.massBucket*iJacBucket(1:2,:) + iJacBucket(3,:)'*obj.c.moiBucket*iJacBucket(3,:);

            % Coriolis and centrifugal terms
            b = iJacBoom(1:2,:)'*obj.c.massBoom*iJacBoomDot(1:2,:)*qDot + ...
                iJacBoom(3,:)'*(obj.c.moiBoom*iJacBoomDot(3,:)*qDot) + ...
                iJacArm(1:2,:)'*obj.c.massArm*iJacArmDot(1:2,:)*qDot + ...
                iJacArm(3,:)'*(obj.c.moiArm*iJacArmDot(3,:)*qDot) + ...
                iJacBucket(1:2,:)'*obj.c.massBucket*iJacBucketDot(1:2,:)*qDot + ...
                iJacBucket(3,:)'*(obj.c.moiBucket*iJacBucketDot(3,:)*qDot);

            % Gravitational terms
            g = -iJacBoom(1:2,:)'*obj.c.massBoom*obj.c.g(1:2) - ...
                iJacArm(1:2,:)'*obj.c.massArm*obj.c.g(1:2) - ...
                iJacBucket(1:2,:)'*obj.c.massBucket*obj.c.g(1:2);

            % External force
            extF = iJacM(1:2,:)'*F;
            
            T = M*qDDot + b + g - extF;
            TBoom = T(1);
            TArm = T(2);
            TBucket = T(3);
        end
        
        %{
        Calculate boom, arm and bucket actuator forces
        Input: Boom, arm and bucket virtual joint torques
        Output: Boom, arm and bucket actuator forces
        %}
        function [FBoom, FArm, FBucket] = calcForces(obj, TBoom, TArm, TBucket)
            % Boom
            angBDC = acos((obj.lenBoom^2 + obj.c.lenBD^2 - obj.c.lenBC^2)/...
                (2*obj.lenBoom*obj.c.lenBD));
            theta = pi/2 - angBDC; % Angle between tangential velocity vector of D and linear velocity of boom actuator
            FBoom = TBoom/(obj.c.lenBD*cos(theta));

            % Arm
            angAFE = acos((obj.lenArm^2 + obj.c.lenAF^2 - obj.c.lenAE^2)/...
                (2*obj.lenArm*obj.c.lenAF));
            theta = angAFE - pi/2; % Angle between tangential velocity vector of F and linear velocity of arm actuator
            FArm = TArm/(obj.c.lenAF*cos(theta));

            % Bucket
            FBucket = TBucket/-(0.0192*obj.ang3^3 + 0.0864*obj.ang3^2 + 0.045*obj.ang3 - 0.1695);
        end

        function visualise(obj)
            hold on;
            axis equal;
            
            % Boom (0.77 - 1.35 m)
            iBA = rotMat(obj.ang1)*obj.c.bBA;
            iBD = rotMat(obj.ang1)*obj.c.bBD;
            iBE = rotMat(obj.ang1)*obj.c.bBE;
            boom = polyshape([[0; 0], iBD, iBA, iBE]');
            
            plot(boom, FaceColor="k");
            plot([0, obj.c.iBC(1)], [0, obj.c.iBC(2)], Color="k"); % BC
            plot([obj.c.iBC(1), iBD(1)], [obj.c.iBC(2), iBD(2)], Color="r"); % Boom actuator
            
            % Arm (1.01 - 1.59 m)
            iBF = transMat(obj.ang1 + obj.ang2, iBA)*[obj.c.aAF; 1];
            iBF = iBF(1:2);
            iBG = transMat(obj.ang1 + obj.ang2, iBA)*[obj.c.aAG; 1];
            iBG = iBG(1:2);
            iBJ = transMat(obj.ang1 + obj.ang2, iBA)*[obj.c.aAJ; 1];
            iBJ = iBJ(1:2);
            iBL = transMat(obj.ang1 + obj.ang2, iBA)*[obj.c.aAL; 1];
            iBL = iBL(1:2);
            
            
            R = 2*obj.c.lenJG; % R formula (sin version)
            theta = atan2(obj.c.aJG(1), obj.c.aJG(2));
            angLJH = asin((-obj.lenBucket^2 + obj.c.lenHJ^2 + obj.c.lenJG^2) ...
                /(R*obj.c.lenHJ)) - theta;
            aAH = transMat(angLJH, obj.c.aAJ)*[obj.c.lenHJ; 0; 1];
            aAH = aAH(1:2);
            iBH = transMat(obj.ang1 + obj.ang2, iBA)*[aAH; 1];
            iBH = iBH(1:2);
            

            arm = polyshape([iBA, iBJ, iBL, iBG, iBF]');

            plot(arm, FaceColor="k");
            plot([iBE(1), iBF(1)], [iBE(2), iBF(2)], Color="r"); % Arm actuator
            plot([iBG(1), iBH(1)], [iBG(2), iBH(2)], Color="r"); % Bucket actuator
            plot([iBJ(1), iBH(1)], [iBJ(2), iBH(2)], Color="k");
            
            % Bucket (0.83 - 1.26 m)
            iBK = transMat(obj.ang1 + obj.ang2 + obj.ang3, iBL)*[obj.c.lLK; 1];
            iBK = iBK(1:2);
            iBM = transMat(obj.ang1 + obj.ang2 + obj.ang3, iBL)*[obj.c.lLM; 1];
            iBM = iBM(1:2);
            bucket = polyshape([iBL, iBM, iBK]');

            plot(bucket, FaceColor="k");
            plot([iBH(1), iBK(1)], [iBH(2), iBK(2)], Color="k");

            % Ground
            yline(obj.c.yGround, Color="g");

            hold off;
        end
    end
end